google-cloud-spanner/src/test/java/com/google/cloud/spanner/it/ITJsonWriteReadTest.java

```diff
@@ -0,0 +1,150 @@
+
+@Category(ParallelIntegrationTest.class)
+@RunWith(JUnit4.class)
+public class ITJsonWriteReadTest {
+
+  private static final String RESOURCES_DIR = "com/google/cloud/spanner/it/";
+  private static final String VALID_JSON_DIR = "valid";
+  private static final String INVALID_JSON_DIR = "invalid";
+
+  private static final String TABLE_NAME = "TestTable";
+
+  @ClassRule public static IntegrationTestEnv env = new IntegrationTestEnv();
+
+  private static DatabaseClient databaseClient;
+
+  @BeforeClass
+  public static void beforeClass() {
+    final RemoteSpannerHelper testHelper = env.getTestHelper();
+    if (!EmulatorSpannerHelper.isUsingEmulator()) {
+      final Database database =
+          testHelper.createTestDatabase(
+              "CREATE TABLE "
+                  + TABLE_NAME
+                  + "("
+                  + "Id INT64 NOT NULL,"
+                  + "json JSON"
+                  + ") PRIMARY KEY (Id)");
+      databaseClient = testHelper.getDatabaseClient(database);
+    }
+  }
+
+  @Test
+  public void testWriteValidJsonValues() throws IOException {
+    assumeFalse("Emulator does not yet support JSON", EmulatorSpannerHelper.isUsingEmulator());
+
+    List<String> resources = getJsonFilePaths(RESOURCES_DIR + File.separator + VALID_JSON_DIR);
+
+    long id = 0L;
+    List<Mutation> mutations = new ArrayList<>();
+    for (String resource : resources) {
+      String jsonStr =
+          Resources.toString(
+              Resources.getResource(this.getClass(), VALID_JSON_DIR + File.separator + resource),
+              StandardCharsets.UTF_8);
+      Mutation mutation =
+          Mutation.newInsertBuilder(TABLE_NAME)
+              .set("Id")
+              .to(id++)
+              .set("json")
+              .to(Value.json(jsonStr))
+              .build();
+      mutations.add(mutation);
+    }
+    databaseClient.write(mutations);
+
+    try (ResultSet resultSet =
+        databaseClient
+            .singleUse()
+            .executeQuery(Statement.of("SELECT COUNT(*) FROM " + TABLE_NAME))) {
+      resultSet.next();
+      assertEquals(resultSet.getLong(0), resources.size());
+    }
+  }
+
+  @Test
+  public void testWriteAndReadInvalidJsonValues() throws IOException {
+    assumeFalse("Emulator does not yet support JSON", EmulatorSpannerHelper.isUsingEmulator());
+
+    List<String> resources = getJsonFilePaths(RESOURCES_DIR + File.separator + INVALID_JSON_DIR);
+
+    for (String resource : resources) {
+      String jsonStr =
+          Resources.toString(
+              Resources.getResource(this.getClass(), INVALID_JSON_DIR + File.separator + resource),
+              StandardCharsets.UTF_8);
+
+      assertThrows(
+          SpannerException.class,
+          () ->
+              databaseClient.write(
+                  Collections.singletonList(
+                      Mutation.newInsertBuilder(TABLE_NAME)
+                          .set("Id")
+                          .to(100L)
```

Comment:
nit: This should use the same kind of increasing id (or other unique id) as in the other test. Otherwise, if two or more invalid JSON values are accepted by the backend, you will get a wrong error message for the second one, as it will fail with a PRIMARY KEY constraint violation instead (in other words: if 10 values all of a sudden are accepted, each test run will only detect the first one)